plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataBuilder.java

```diff
@@ -0,0 +1,356 @@
+    String release = latest;
+    while (release.contains("SNAPSHOT") && vi.hasNext()) {
+      release = vi.next().toString();
+    }
+    if (release.contains("SNAPSHOT")) {
+      release = null;
+    }
+    return Maven2Metadata.newArtifactLevel(
+        DateTime.now(),
+        getGroupId(),
+        getArtifactId(),
+        latest,
+        release,
+        Iterables.transform(baseVersions, new Function<Version, String>()
+        {
+          @Override
+          public String apply(final Version input) {
+            return input.toString();
+          }
+        }));
+  }
+
+  private void addBaseVersion(final String baseVersion) {
+    checkNotNull(baseVersion, "baseVersion");
+    try {
+      if (baseVersions.add(versionScheme.parseVersion(baseVersion))) {
+        log.debug("Added base version {}:{}:{}", getGroupId(), getArtifactId(), baseVersion);
+      }
+    }
+    catch (InvalidVersionSpecificationException e) {
+      // According to versionScheme implementation we use, this exception will never happen
+      // log + ignore
+      log.info("Invalid baseVersion discovered: " + baseVersion , e);
+    }
+  }
+
+  // -----------------------------------
+  // baseVersion
+
+  /**
+   * Internal structure to hold parsed Aether {@link Version} and {@link Coordinates}.
+   */
+  private static class VersionCoordinates
+  {
+    private final Version version;
+
+    private final Coordinates coordinates;
+
+    private VersionCoordinates(final Version version, final Coordinates coordinates) {
+      this.version = version;
+      this.coordinates = coordinates;
+    }
+  }
+
+  public boolean onEnterBaseVersion(final String baseVersion) {
+    if (setBaseVersion(baseVersion)) {
+      log.debug("-> GAbV: {}:{}:{}", getGroupId(), getArtifactId(), getBaseVersion());
+      latestVersionCoordinatesMap.clear();
+      latestVersionCoordinates = null;
+      return true;
+    }
+    return false;
+  }
+
+  @Nullable
+  public Maven2Metadata onExitBaseVersion() {
+    checkState(getBaseVersion() != null, "baseVersion");
+    log.debug("<- GAbV: {}:{}:{}", getGroupId(), getArtifactId(), getBaseVersion());
+    if (!getBaseVersion().endsWith("SNAPSHOT") || latestVersionCoordinates == null) {
+      // release version does not have version-level metadata
+      log.debug("Not a snapshot or nothing to generate: {}:{}:{}", getGroupId(), getArtifactId(), getBaseVersion());
+      return null;
+    }
+    final List<Snapshot> snapshots = Lists.newArrayList();
+    for (VersionCoordinates versionCoordinates : latestVersionCoordinatesMap.values()) {
+      final Coordinates coordinates = versionCoordinates.coordinates;
+      final Snapshot snapshotVersion = Maven2Metadata.newSnapshot(
+          new DateTime(coordinates.getTimestamp()),
+          coordinates.getExtension(),
+          coordinates.getClassifier(),
+          coordinates.getVersion()
+      );
+      snapshots.add(snapshotVersion);
+    }
```

Comment:
minor: clearing some of the accumulated data on exit (rather than only on enter) would help keep memory usage down